### ISSUE TITLE
Deepcopy doesn't copy to same gpu

### ISSUE THREAD

**Issue by crazyoscarchang** [pytorch/pytorch#14673]
This issue was first reported here: https://discuss.pytorch.org/t/deepcopy-with-multiple-gpus/20283

```
from copy import deepcopy
import torch

x = torch.ones((1,), device=torch.device("cuda", 1))
print(x)
## result : tensor([ 1.], device='cuda:1') 
y = deepcopy(x)
print(y)
## result : tensor([ 1.], device='cuda:0') 
```
This code will copy x onto the first GPU with cuda id 0, instead of on the original GPU with cuda id 1.

deepcopy is necessary when we want to avoid autograd tracking gradients into the new copy, unlike .clone()

**Comment by yaox12** [comment]
I ran into the same problem when I tried to deepcopy a module.
I'm now working around by adding a `.to(device=some_device)` after deepcopy

**Comment by zou3519** [comment]
Thank you for surfacing this, @crazyoscarchang, we'll get this fixed soon.

**Comment by vishwakftw** [comment]
I've found out the issue. Cloning `torch.Storage` doesn't respect GPU ids, and clone to the default GPU on which the context is created. I'll try to fix this, and send in a patch as soon as possible.

**Comment by soumith** [comment]
@vishwakftw someone is prepping a patch already

**Comment by vishwakftw** [comment]
Oh okay, no worries. Thank you for letting me know.
